### DISCUSSION TITLE
Smart Home watchapp

### DISCUSSION THREAD

**Discussion by Raupinger** [InfiniTimeOrg/InfiniTime#1608] [Ideas]
This is probably something to work on down the line, but an App to control smart lighting via openhab would be nice to have. This would obviously need support in the companion apps which doesn't currently exist if I'm not mistaken.

**Comment by JF002** [comment]
Smart home application was one of my original project! But building a whole application firmware deverted me from this goal :)

As PineTime can only communicate on BLE, you'll indeed need a companion app or a bridge device to bridge your pinetime to your smart home application (OpenHAB or Home-assistant in my case).

**Reply by weeowey** [reply]
BTHome.io could be perfect

**Comment by pfeerick** [comment]
Sounds like a perfect job (bridge device) for the pinecone! ;)

**Comment by sporkbyspork** [comment]
For the glorious Pebble, there are companion apps like [PebbleTasker](https://play.google.com/store/apps/details?id=com.kodek.pebbletasker&hl=en&gl=US) or
[AutoApp for Pebble](https://play.google.com/store/apps/details?id=com.joaomgcd.autopebble&hl=en&gl=US).
You create the situation-aware menus on your phone and the watch shows them. Then you can use the almighty Tasker to react.

I dread the moment when my Pebble dies and there are no more replacements. So a comparable functionality would be a great way to get people to jump onto the PineTime bandwagon.

**Comment by funnyflowerpot** [comment]
I would also be very much in favour of possibilities to add smart home capabilities to the PineTime!

This might be a bit of a tangent, but I wanted to share experiences briefly. In [a comment](https://github.com/InfiniTimeOrg/InfiniTime/issues/562#issuecomment-907613478) in a related issue @unrelentingtech was linking to a [project on Github](https://github.com/shmuelzon/esp32-ble2mqtt/) that implements a BLE/GATT to WiFi/MQTT bridge on an ESP32. I found the project a while ago, tried it out, and connected the PineTime to it. 

Summary: I was able to send notifications to my watch via MQTT with for example the shell script below. It worked quite well and liked it a lot! 

```
PINETIME_MAC="11:22:33:44:55:66"
MQTT_TOPIC="$PINETIME_MAC/AlertNotificationService/NewAlert/Set"

printf '\x00\x01%s\x00%s' "$1" "$2" | \
  od -tu1 -An | xargs | tr ' ' ',' | \
  mosquitto_pub -t "$MQTT_TOPIC" -s
```

To me this was an eye-opener for so many new possibilities and I would really love to see development into this direction! Maybe I can get into InfiniTime development later this year.

Also, as a more general observation, I noted that a BLE-MQTT-bridge (or anything similar) can remove the need of using a smartphone or companion app, which is a great advantage in certain situations. This of course depends on personal use case. However, it could be seen as trading dependency on the phone for location-based functionality (if the ESP32/etc. is permanently installed). This would be an advantage for some people (including me, who would love to get rid of wearing the smartphone all the time in favour of the PineTime).

**Comment by cpyarger** [comment]
It would be nice to see this as something that can run as an addon via home assistant Something like an extension of the Passive BLE monitor integration,

**Comment by cpyarger** [comment]
I have submitted an issue with a new sensor request, in order to have the infinitime watch added as a supported sensor with the passive ble monitor integration. Maybe if there is enough chatter on the issue, it may be considered. [New Infinitime Sensor](https://github.com/custom-components/ble_monitor/issues/846)

**Comment by JF002** [comment]
@cpyarger  I'm a user of Home Assistant, but I don't know about that passive BLE monitor Integration. If I understand correctly, this would allow to detect the PineTime and even poll for step or HR data and use those data in HA?

**Comment by cpyarger** [comment]
> @cpyarger I'm a user of Home Assistant, but I don't know about that passive BLE monitor Integration. If I understand correctly, this would allow to detect the PineTime and even poll for step or HR data and use those data in HA?

Yes. That is the idea. Also setting the time. And acting as a endpoint for notifications

**Comment by JF002** [comment]
That's really nice ! I hope someone will consider working on it, then!

**Comment by bbadrignans** [comment]
I made a prototype app that is able to communicate with Home Assistant :
https://community.home-assistant.io/t/pinetime-infinitime-gadgetbrdige-support/550305
Is anyone interested by such app ?

**Comment by Aquariu** [comment]
Hello everyone,
My name is Olivier and I own both a Fossil Q Hybrid (the old version)  and most recently got a Pine Time (running Infinitime 1.11 for now).
I was looking at a "better" way to control some automation than  what I had with the Fossil...

On the Fossil Q, gadgetbridge registers an Intent when I press one of the smartwatch's buttons that I have to grab with Tasker to then do something with it. In Domoticz, Jeedom or other open source self-hostable home automation systems, you can usually perform an action with just a URL. So my Tasker script would call the URL with the correct API key and it would sort of work.
Alas, performance is not good in that when the watch is asleep, the BLE connection would sometime fail and the button press would not register. OR the phone being asleep would miss it. Either way, it didn't work more often than not. And my garage door management was a bit chaotic. (it's described [here](https://codeberg.org/Freeyourgadget/Gadgetbridge/issues/965) if that is of interest for some of you, but beware, it's a looong thread with the juicy info [right in the middle](https://codeberg.org/Freeyourgadget/Gadgetbridge/issues/965#issuecomment-40110))

I hoped to find a better solution in the Pine Time app or a wake-up scheme.
Reading that discussion and the various Issues that google lead me to. It looks like
- There are so many automation systems that it looks difficult to have one app to rule them all when you can just grab the intent and 
- An app with just 4 squares with numbers, and leaving the companion app doing the heavy lifting (and passing the ball to the triggering mechanism) looks like it's the easiest solution from the watch point of view. 


There's one capability that may justify doing something else but from the discussion above it looks like it's a heavy firmware modification : many automation systems can directly use BLE beacons to listen to events. So instead of doing something convoluted like `Watch--> BLE --> Phone --> HTTP/MQTT --> automation web server`,  we would just  have `Watch --> BLE --> Automation BLE beacon`.  currently, I can use old Nodon Niu bluetooth buttons to open my garage door. Works great. If I could configure an app or just a special wake-up (triple tap, double button press, whatever... ) that would be great.

Sorry for the long message, I'm new here and not sure about how to go about it.

**Reply by alphasixtyfive** [reply]
We would still need `Watch--> BLE --> Phone --> HTTP/MQTT --> automation web server` route when being away from home. Also, BLE and large homes don't work together that great. I'm now playing with HomeAssistant and ESP Home BT relays scattered around the house but it is not as reliable as I'd like it to be.

**Reply by bbadrignans** [reply]
@Aquariu you can look here : https://community.home-assistant.io/t/pinetime-infinitime-gadgetbrdige-support/550305
It works like this : 
`Watch --> BLE --> Gadget Bridge on phone --(intent)--> Home assistant App on phone --> Home assistant server`
This avoid using Tasker in the middle. The persistent connexion from Home assistant app to the server is pretty reliable.
I use it daily for months and it works 99% of time. 
However it requires modifications on Gadget Bridge and pinetime firmware that I did not push to the main branch yet.

**Reply by alphasixtyfive** [reply]
I'm really looking forward to seeing this as part of the official firmware!

**Comment by Aquariu** [comment]
> We would still need `Watch--> BLE --> Phone --> HTTP/MQTT --> automation web server` route when being away from home. Also, BLE and large homes don't work together that great. I'm now playing with HomeAssistant and ESP Home BT relays scattered around the house but it is not as reliable as I'd like it to be.

Indeed. If I need to trigger things at home, i most often use the web interface or the phone. So I hadn't really thought about using the watch when away. But yeah I guess :) Why not !

**Reply by alphasixtyfive** [reply]
For me it is mostly opening the gate when I'm coming close to the house.

**Comment by agittins** [comment]
An (opinionated) update on where things are at in Homeassistant + Bluetooth land...

- Using esp32 based esphome units as [bluetooth proxies](https://esphome.io/components/bluetooth_proxy.html) in HA works quite well, and gives full-house, fault-tolerant coverage for receiving BLE adverts, and offers outbound connections.
- Shelly Gen2 devices can also act as native BLE advertisement proxies.
- The [BTHome](https://bthome.io/) protocol offers a way to send sensor and event data directly to HA without having to create an integration or write any code at the other end. If you broadcast bthome packets, HA will discover them, offer the user to create a device for it, and create sensors etc. BTHome also supports pre-shared key encryption.
- I have not used the bluetooth proxies for actually connecting *to* devices (I am using them just for reception of BLE broadcasts currently) but there are a number of integrations that use it and seem to work, providing connection sessions are kept short.

I currently use my Pinetime running infinitime as a BLE "beacon", tracking it with an integration I wrote called Bermuda, which gives me room-based occupancy sensing. I patched Infinitime to reduce the non-fast advertising interval to around 0.5sec for better latency/detection. It works quite well, but I suspect it's using a lower power level but I haven't dug into it.

I see a (relatively) easy path to being able to use Infinitime with Homeassistant, if anyone has the appetite to take it on:

Create an infinitime app that presents the user with (n) screens containing 4 buttons. Pressing one of the buttons sends a BTHome broadcast packet (perhaps a blast of several, with the same packet_id) with `event` objects to notify the state of the buttons. This, with no extra work in HA, will create a "device" with 4 buttons that can then be used for HA-side automations.

Secondarily, an integration in HomeAssistant could be created as an "Infinitime" custom component. It could provide any level of the infinitme bluetooth protocol.
- Being able to send call notifications and take responses would allow automation prompts like "Garage door has been open for two hours, press Answer to close it, Mute to ignore for an hour" etc
- A new message format (or something) could be created which tells the watch to change the icons on the BTHome buttons and/or change the "title" of a given page of buttons. This way HA could create automations that update the watch's BTHome buttons so that (for example) if you enter a new room, the watch's buttons would change to match devices available in that room: like a fan, light, and curtains for a bedroom, and when you enter the garage it changes them to lights, garage door, etc.

This would provide a huge amount of flexibility for users, as they could update and customise what the watch can do without having to compile and flash new firmware, instead just building automation scripts to act on the buttons differently based on whatever state they like, plus if an integration were written, update the icons or labels, as well as do other tasks like notifications, firmware updates, weather updates, media controls for anything in the home, etc etc.

I think Infinitime is probably the best-placed to do this efficiently. Personally, I am going to have a crack at it via Wasp-OS though, as it's a much more comfortable environment for me, but it would be an amazing setup to see in Infinitime!

Another option is of course having HA talk to the HA Android app, with sends intents to Gadgetbridge, which sends messages to/from Infinitime, but it's a lot of bouncing around and a complex stack to guide users through setting up, and also a rather rigid framework that gets in the way of customisation, and for me I need Infinitime to be broadcasting so I don't have my phone connected to it most of the time. Upside to going via GB is that it would work outside the home as well, but a BTHome app would still be useful for local control.

**Reply by weeowey** [reply]
I want a BTHome app too! I use it on my flipper zero all the time to trigger automations when coming home. Imagine we could sync our health data automatically to HA and just easily look at trends, etc. We could have multiple buttons to tap, hold or double/triple press.. Game changer even if it is just going to be local. We could even suggest the route of infinitime just having a straight up BTHome mode where it just acts as a beacon for other automations and an app for button presses. I don't like having my watch connected to my phone, plus the battery life is much better when not connected too.

**Comment by BeatLink** [comment]
I am also very interested in this feature. I think the best way forward is a generic remote control app that sends events to a companion app, then have the companion app include code to call other services as needed, maybe even using Tasker as an inbetween. 

The pros of this is the companion app would do the heavy lifting, all the watch would need is to show a couple of numbered buttons. Also it would not just tied to home assistant alone but can be used in a variety of other use cases (like a speed dial button that calls favorite contacts)
